### ISSUE TITLE
Upgrading Cake and Git Strategy

### ISSUE THREAD

**Issue by jeffreyroberts** [cakephp/cakephp#10715]
This is a (multiple allowed):

* [x] bug
* [x] enhancement
* [x] feature-discussion (RFC)

* CakePHP Version: All three versions
* Platform and Target: LAMP

### What you did
EXPLAIN WHAT YOU DID, PREFERABLY WITH CODE EXAMPLES, HERE.

git clone https://github.com/cakephp/cakephp.git MyProject
cd MyProject
rm -rf .git
git init
git remote add origin git@myhosting.none:user/myproject.git
git add . && git commit -m 'Initial Shotgun Commit' && git push origin master

### What happened
EXPLAIN WHAT IS ACTUALLY HAPPENING, HERE.

What I am doing is setting up my next project, and im wondering why if I want to upgrade cakephp from 2.9.8 to 2.9.9 then I have to deal with multiple remotes, it seems like to me the easiest way to do this, would to be to have a directory structure that would fit into a usr/local (or lib with lib/bin) folder of a parent repo and I know the directory structure has been discussed thoroughly but I think you guys are missing something with git in a couple of places, I would really like to produce a report on why I think the directory structure should be updated to a more upgrade friendly git strategy because right now, its not very friendly...

### What you expected to happen
EXPLAIN WHAT IS TO BE EXPECTED, HERE.

mkdir MyProject
git init
git submodule add https://github.com/cakephp/cakephp.git lib

... and to upgrade ...

cd MyProject/lib
git pull

So basically, I am wondering who I can submit a report and possible solution to about this issue, I dont think a pull request is going to do it, I know a lot of thought has gone into the current and previous structures, I would just like to try to impose myself into that debate a bit to bring up git and upgrade strategy...

**Comment by markstory** [comment]
Have you considered [using composer](https://book.cakephp.org/2.0/en/installation/advanced-installation.html#installing-cakephp-with-composer)? That is the way I would recommend installing CakePHP. I agree that the current 2.x directory layout in the repository that combines the app and framework is not ideal, but we are not going to change it for 2.x as it would break the existing workflows that people have.

**Comment by jeffreyroberts** [comment]
Ive looked into composer a bit, not quite fluent yet though...

I am a huge fan of Cake, and Plugins. I have done a lot of work extending plugin functionality, nested plugins for one, and I have to say, I love it all very much... 

Have you used https://github.com/basecamp/sub before?

After you clone it, you run a script, I suggest you give it a clone and try on the install... I think an approach like the install script would be amazing with cake...

I would love to imagine a cake one day that would allow me to bake an app, and cake would add it as a submodule for me, then I can just add the remote and push, and what would even be better is a post-bake-app.d script directory or something like that where I can script out adding my remote to the submodule...

I have a ton of ideas (feature requests), I have been working with Cake for many years now, and I would love to contribute, do you guys have dev wiki for notes and thoughts where I can bang around some ideas before going gungho into coding?

**Comment by jeffreyroberts** [comment]
ok, so I guess what I am saying is, is maybe automating a little bit of git with cake, multiple project battle testing, might bring out the best folder structure, just some food for thought

**Comment by jeremyharris** [comment]
@jeffreyroberts 

Glad to hear you want to contribute! Usually, if you're unsure about whether or not a feature/idea would be a good fit for Cake, you would open an RFC issue ticket here on Github to start a discussion. You can provide pseudo code if you like, however it's not necessary. If there's a consensus that it would be a good fit for the core, then you would do some coding and open an actual PR.

You can also start by creating the functionality in the form of a plugin and submit it to https://github.com/FriendsOfCake/awesome-cakephp.

As for simple installation, I would agree with Mark and highly suggest you look into composer. I was a die-hard git submodule guy for a while but once I learned how to properly use composer I have never looked back. You can also bundle the phar with your app and use that if it's not (or you can't) install it on your host.

As for your wish to simply push a git repo and have a deploy happen, there are several PaaSes out there that make that super easy. (I like Pagoda Box, for example. Hit me up on slack and I can help get you started.) They even let you do pre/post deploy/build hooks.

Using a PaaS and composer, your upgrade workflow would be a simple as:

```
$ composer update cakephp/cakephp
$ git add composer.json composer.lock
$ git commit -m "Upgrade cake"
$ git push production master
```

**Comment by ADmad** [comment]
@jeffryroberts  The folder structure for cakephp apps has already been changed in 3.x and composer is the officially recommended way to handle package updates. In 3.x the cakephp core (and even plugins) is installed as a vendor and can be easily upgraded without​ affecting your app code. So please check the docs and start using cake 3.

We cannot afford to spend time and effort changing the folder structure of cake 2.x.

**Comment by markstory** [comment]
CakePHP 3 also has a simple install script via [Oven](https://github.com/CakeDC/oven) which helps automate getting a project setup with composer.

**Comment by jeffreyroberts** [comment]
rock and roll @ Oven, I think I need to go back over the psr-7 standard and catch up with 3.x before I comment further, I was having a 2.x moment, you guys are doing a great job, keep up the amazing work =]

**Comment by jeffreyroberts** [comment]
let me put it this way, if I can upgrade via git pull, then I can contribute via git push... sounds like to me, composer needs to use git, especially if its defacto these days, imagine how many lost contributions there are because people cant just git push fixes

**Comment by jeremyharris** [comment]
Composer usually uses git, using the tag feature of git to pull the specific sha. Contributing to a package is as easy as a PR. When it is merged, it will be included in the next tag.

Composer is the standard for dealing with php dependencies, for which it uses git to actually pull the repos down. It's like submodules with dependency management. And much more.

P.S. I think this discussion would be better suited for the forum 😊

**Comment by ADmad** [comment]
@jeffreyroberts We appreciate your feedback but I suggested you familiarize yourself more first with composer and current structure of CakePHP 3.x apps.

Discussions like this are better suited for the forum.

Closing the ticket as there's actually nothing to be done here.

**Comment by jeffreyroberts** [comment]
Sorry about that, I will locate the forum asap, sounds like I need some educating
